client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java

```diff
@@ -63,15 +63,22 @@ private FixJsonDataUtils() {}
             return null;
         }
         requireNonNull(columns, "columns is null");
-        List<ClientTypeSignature> signatures = columns.stream()
+        ClientTypeSignature[] signatures = columns.stream()
                 .map(Column::getTypeSignature)
-                .collect(toList());
-        ImmutableList.Builder<List<Object>> rows = ImmutableList.builder();
+                .toArray(ClientTypeSignature[]::new);
+        ImmutableList.Builder<List<Object>> rows;
+        if (data instanceof Collection<?>) {
+            rows = ImmutableList.builderWithExpectedSize(((Collection<?>) data).size());
+        }
+        else {
+            rows = ImmutableList.builder();
+        }
         for (List<Object> row : data) {
-            checkArgument(row.size() == columns.size(), "row/column size mismatch");
-            List<Object> newRow = new ArrayList<>();
-            for (int i = 0; i < row.size(); i++) {
-                newRow.add(fixValue(signatures.get(i), row.get(i)));
+            int rowColumns = row.size();
```

Comment:
Does extracting this variable actually matters? (it definitely impacts readability)

Reply:
Probably not, will leave it as `row.size()`